processor/src/memory/mod.rs

```diff
@@ -224,12 +224,44 @@ impl Memory {
         self.get_values(RangeInclusive::new(0, u64::MAX))
     }
 
-    /// Returns values within a range of addresses.
+    /// Returns values within a range of addresses at the last clock cycle.
     pub fn get_values(&self, range: RangeInclusive<u64>) -> Vec<(u64, Word)> {
+        self.get_values_at(range, None)
+    }
+
+    /// Returns values within a range of addresses, or optionally all values at a partciular step.
+    /// If no step is specified, then values at the last step are returned.
+    pub fn get_values_at(
+        &self,
+        range: RangeInclusive<u64>,
+        step: Option<usize>,
+    ) -> Vec<(u64, Word)> {
         let mut data: Vec<(u64, Word)> = Vec::new();
-        for (&addr, addr_trace) in self.trace.range(range) {
-            let value = addr_trace.last().expect("empty address trace").1;
-            data.push((addr, value));
+        if let Some(s) = step {
+            for (&addr, addr_trace) in self.trace.range(range) {
+                match addr_trace.binary_search_by(|(x, _)| x.as_int().cmp(&(s as u64))) {
+                    Ok(i) => data.push((addr, addr_trace[i].1)),
+                    Err(mut i) => {
+                        // binary search would find the index that step should be in,
+                        // or the index that we should insert in.
+                        // We decrement the index to find the largest step before that index to find
+                        // the last step available in that address.
+                        // However, if the index is zero then we just pick the first element
+                        // to avoid going out of bounds.
+                        if i > 0 {
+                            i -= 1;
+                        }
+                        if addr_trace[i].0.as_int() <= s as u64 {
+                            data.push((addr, addr_trace[i].1));
+                        }
```

Comment:
if the index is 0, would it mean that clock cycle of the first element is greater than `s`? If so, would the following code do the same thing?:

```Rust
if i > 0 {
  data.push((addr, addr_trace[i - 1].1));
}
```

Reply:
I believe it also returns zero when there is only one element in the vec, or when it should be inserted into the front.
I think we still want to push in the latter case, that's why I make sure it doesn't go below 0 but also still go check if we should do the push.

Reply:
yes, but if `i==0` it means `addr_trace[i].0 > s`, right? otherwise we'd be in the `Ok` branch. Or am I missing something?

Reply:
Ah that's right, I fixed it now